[bundlejs] add badge for the npm package size

crimx opened the pull request:
View [esbuild](https://github.com/evanw/esbuild) minified and [denoflate](https://github.com/hazae41/denoflate) gzipped size of a package or selected exports, via [bundlejs](https://bundlejs.com/).

Comparing to the existing bundlephobia service, bundlejs is more accurate in terms of treeshaking. It also supports displaying only the size of the selected exports.

cc @okikio

okikio commented:
@crimx Thanks

chris48s commented on services/bundlejs/bundlejs-package.tester.js:
```diff
@@ -0,0 +1,33 @@
+export const t = await createServiceTester()
+
+t.create('bundlejs/package (packageName)')
+  .timeout(20000)
```
Before reviewing this in any detail, seeing these timeouts made me want to look into API peformance.

For shields we really need the upstream API call to usually return in under 4 seconds for a badge to be viable. GitHub (which is the place where badges are most commonly viewed) has a hard limit of 4 seconds on an image request so if it takes us >4 seconds to render a badge, they just don't get displayed.

Having tried making a few API calls, it seems sometimes the request does take a really long time (23 seconds was the longest I observed), but then the second call to the same URL is quick. I guess the API is slow if the request is a cache miss but then fast once the data is cached.

@okikio - are you also to provide any info on this? Also there any rate limits on using this API, etc?

okikio replied:
That's due to the bundlejs API bundling the package 📦. 

By default bundlejs caches the bundle on request but we set the TTL to 1 day to allow for new versions to propagate quickly.

chris48s replied:
This feels like a really similar issue to what we had with socket.dev in https://github.com/badges/shields/pull/7727

For a really popular package its probably not a huge issue if one request in 24 hours is really slow. For smaller less popular packages, its probably more of an issue. I don't think I have a really clear answer right now, but it might be a good one for us to discuss next time we catch up @calebcartwright

okikio replied:
For smaller packages it's usually rather fast, it the larger packages that are the largest slow down

crimx replied:
Theoretically bundlejs should be faster than bundlephobia who is using webpack. @okikio Have you looked into the performance bottleneck of the service?

okikio replied:
Bundlephobia is bundling all the packages and storing them permanently, I'm currently not doing that because bundlejs allows for custom code so you can get treeshaken bundle sizes, and I just hadn't gotten around to setting up a permanent database for storing all the bundle size data for simple straight forward bundles, e.g. react@18.0.0

I'll get on it as soon as I can

crimx replied:
Bundlephobia also caches tree-shaking size of each export. They are limited and inaccurate most of the time due to the webpack config.

okikio replied:
Yeah, in the future I hope to also do treeshake bundle analysis for all the exports in a package, right now you have to manually give it a the functions and methods you'd like tteeshaken

chris48s replied:
Hello. After some discussion in https://github.com/badges/shields/issues/9174 we've settled on a pattern for handling services that work in this way and implemented it in https://github.com/badges/shields/pull/9233

I appreciate this has take a while, but would you be able to rebase this PR on master and update it use the approach shown in code example in https://github.com/badges/shields/pull/9233 ?

crimx replied:
Sure.

okikio replied:
Yeah, don't use `no-cache` if you can avoid it

crimx replied:
> Yeah, don't use `no-cache` if you can avoid it

Any ETA for the fix? Last time I checked the package size was still incorrect. Maybe cache with `q` plus other search params?

okikio replied:
I forgot to leave a message here this has now been fixed

calebcartwright replied:
> bundlejs now permanently caches the bundle size for packages

Does this mean that the timeout behavior will only ever occur for first-time packages, and that e.g. the size for jquery is permanently stored? If so then we could remove these timeouts from the code (it'll never get that far anyway given the timeout short circuit in the request fetching)

chris48s replied:
Yeah I think given this badge will always return a response of some description (even if that is `bundling`) within 3 seconds, we should be able to remove all these `.timeout(20000)`s .

calebcartwright commented:
Just a bit of a triage ping. Is there still an intent to overhaul this to incorporate some of the subsequent changes that landed on our default branch, including the new pattern Chris mentioned in https://github.com/badges/shields/pull/9055#discussion_r1228651419 that's specifically geared for upstream services/platforms that exhibit this type of behavior?

If not perhaps we could close this and shift discussion back to #8266 in case someone else is willing/interested in working on it?

crimx commented:
I was waiting for the [`options.timeout.request` API](https://github.com/badges/shields/issues/9174#issuecomment-1650384154). [@okikio](https://github.com/okikio) just metioned that the bundlejs results are permanently cached now. I can also make other workarounds for testing.

crimx commented:
@calebcartwright ready to merge

chris48s commented:
Thanks for updating. I can have a look at reviewing this one over the weekend.

calebcartwright commented on services/bundlejs/bundlejs-package.service.js:
```diff
@@ -0,0 +1,130 @@
+    pattern: ':scope(@[^/]+)?/:packageName/:exports?',
+  }
+
+  static examples = [
+    {
+      title: 'npm package minimized gzipped size',
+      pattern: ':packageName',
+      namedParams: {
+        packageName: 'react',
+      },
+      staticPreview: this.render({ size: '2.94 kB' }),
+      keywords,
+      documentation,
+    },
```
Will defer to Chris on whether we want to let this through as-is and refactor with the others later, or try to apply the new model now.

(we've changed the way we set some of these as part of an overhaul to the frontend site)

chris48s replied:
I'm not really worried about this. Given the documentation still says to use `examples`, I think it is unreasonable to expect contributors to do something else as this point.

calebcartwright reviewed (commented):
Couple comments left below

chris48s commented on services/bundlejs/bundlejs-package.service.js:
```diff
@@ -0,0 +1,130 @@
+
+const schema = Joi.object({
+  size: Joi.object({
+    compressedSize: Joi.string(),
```
```suggestion
    compressedSize: Joi.string().required(),
```

If we're going to use `compressedSize`, it needs to be `.required()`

chris48s commented on services/bundlejs/bundlejs-package.service.js:
```diff
@@ -0,0 +1,130 @@
+export default class BundlejsPackage extends BaseJsonService {
+  static category = 'size'
+
+  static route = {
+    base: 'bundlejs/package',
+    pattern: ':scope(@[^/]+)?/:packageName/:exports?',
+  }
```
Based on our URL conventions https://github.com/badges/shields/blob/master/doc/badge-urls.md , I'd say:

- the base should be `bundlejs/size` - the thing we're actually displaying on the badge is a size, not a package
- `exports` should be a query param, not part of the route. It is optional and I'd expect it to be relatively infrequently used

chris48s commented on services/bundlejs/bundlejs-package.tester.js:
```diff
@@ -0,0 +1,33 @@
+import { isFileSize } from '../test-validators.js'
+import { createServiceTester } from '../tester.js'
+export const t = await createServiceTester()
```
Could we add one test cases that uses `.replyWithError()` https://github.com/nock/nock#replying-with-errors to test the `ETIMEDOUT` case

crimx replied:
Don't know how to test it if the result is permanently cached. Maybe to fake one but I doubt if it's useful.

crimx replied:
Found a way to test it.

crimx replied:
Nope. Still unable to test this.

chris48s commented on services/bundlejs/bundlejs-package.tester.js:
```diff
@@ -0,0 +1,33 @@
+import { isFileSize } from '../test-validators.js'
+import { createServiceTester } from '../tester.js'
+export const t = await createServiceTester()
```
Just for clarity, is the strategy we're relying on for the tests here just picking really popular packages that are likely to be already cached at any given time?

crimx replied:
Bundlejs now [permanently caches the result](https://github.com/badges/shields/pull/9055#discussion_r1292840083).

crimx commented:
@chris48s Moved exports to query params and removed test timeouts.

When requesting non-existing packages, bundlejs responses with 404 after unpkg timeout. The timing is unpredictable so it may also trigger the ETIMEDOUT. I changed the badge message to "timeout" for both.

chris48s commented:
So with the 404 case, we're actually hitting the original behaviour. We hit the timeout first time round, but the second request we get the 404 immediately.

I think what we should actually do here is give the user a real "not found" error in the 404 case (if it comes in under the timeout) and test the 404 and timeout cases using mocks, like this:

```js
t.create('bundlejs/package (not found)')
  .get('/react@18.2.0.json')
  .intercept(nock =>
    nock('https://deno.bundlejs.com')
      .get(/./)
      .query({ q: 'react@18.2.0' })
      .reply(404),
  )
  .expectBadge({ label: 'bundlejs', message: 'package or version not found' })

t.create('bundlejs/package (timeout)')
  .get('/react@18.2.0.json')
  .intercept(nock =>
    nock('https://deno.bundlejs.com')
      .get(/./)
      .query({ q: 'react@18.2.0' })
      .replyWithError({ code: 'ETIMEDOUT' }),
  )
  .expectBadge({ label: 'bundlejs', message: 'timeout' })
```

crimx commented:
Updated. Thanks for the help.

chris48s reviewed (approved):
Sorted. Thanks for getting this one over the line!

okikio commented:
OH, wow, this is amazing, great work @crimx and @chris48s
